astropy/table/tests/test_operations.py

```diff
@@ -961,6 +961,105 @@ def test_mixin_functionality(self, mixin_cols):
+        assert out.pformat() == ['  a [2]      b [2]  ',
+                                 '---------- ---------',
+                                 '0.0 .. 0.0 foo .. --',
+                                 '1.0 .. 1.0 bar .. --']
+
+    def test_stack_basic_outer(self, operation_table_type):
+            if operation_table_type is QTable:
+                pytest.xfail('Quantity columns do not support masking.')
+            self._setup(operation_table_type)
+            t1 = self.t1
+            t2 = self.t2
+            t4 = self.t4
+            t12 = table.cstack([t1, t2], join_type='outer')
+            assert t12.pformat() == ['  a [2]      b [2]     c [2] ',
+                                     '---------- ---------- -------',
+                                     '0.0 .. 2.0 foo .. pez -- .. 4',
+                                     '1.0 .. 3.0 bar .. sez -- .. 5']
+            t124 = table.cstack([t1, t2, t4], join_type='outer')
+            assert t124.pformat() == ['  a [3]      b [3]     c [3]  ',
+                                      '---------- ---------- --------',
+                                      '0.0 .. 0.0 foo .. foo -- .. --',
+                                      '1.0 .. 1.0 bar .. bar -- .. --']
+
+    def test_stack_basic_inner(self, operation_table_type):
+        self._setup(operation_table_type)
+        t1 = self.t1
+        t2 = self.t2
+        t4 = self.t4
+
+        t12 = table.cstack([t1, t2], join_type='inner')
+        assert type(t12) is operation_table_type
+        assert type(t12['a']) is type(t1['a'])
+        assert type(t12['b']) is type(t1['b'])
+        assert t12.pformat() == ['  a [2]      b [2]   ',
+                                 '---------- ----------',
+                                 '0.0 .. 2.0 foo .. pez',
+                                 '1.0 .. 3.0 bar .. sez']
+
+        t124 = table.cstack([t1, t2, t4], join_type='inner')
+        assert type(t124) is operation_table_type
+        assert type(t12['a']) is type(t1['a'])
+        assert type(t12['b']) is type(t1['b'])
+        assert t124.pformat() == ['  a [3]      b [3]   ',
+                                  '---------- ----------',
+                                  '0.0 .. 0.0 foo .. foo',
+                                  '1.0 .. 1.0 bar .. bar']
+
+    def test_multi_dimension_column_stack(self, operation_table_type):
+        self._setup(operation_table_type)
+        t3 = self.t3
+        t5 = self.t5
+        t2 = self.t2
+        t35 = table.cstack([t3, t5])
+        assert type(t35) is operation_table_type
+        assert type(t35['a']) is type(t3['a'])
+        assert type(t35['b']) is type(t3['b'])
+        assert t35.pformat() == [' a [2,2]    b [2,2]  ',
+                                 '---------- ----------',
+                                 '5.0 .. 2.0 foo .. pez',
+                                 '4.0 .. 6.0 pez .. sez']
+
+        with pytest.raises(TableMergeError):
+            table.cstack([t2, t3])
+
```

Comment:
Is it okay that we cannot merge column having a different shape for cstack?

Reply:
Yes. This sort of stacking depends on the input column shapes being the same. The same happens for numpy vstack:
```
In [16]: a = np.array([[1, 2]])

In [17]: b = np.array([[3, 4]])

In [18]: np.vstack([a, b])
Out[18]: 
array([[1, 2],
       [3, 4]])

In [19]: b = np.array([[[3, 33], [4, 44]]])

In [20]: np.vstack([a, b])
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
~/git/astropy/cstack.py in <module>()
----> 1 np.vstack([a, b])

~/miniconda3/envs/astropy/lib/python3.6/site-packages/numpy/core/shape_base.py in vstack(tup)
    232 
    233     """
--> 234     return _nx.concatenate([atleast_2d(_m) for _m in tup], 0)
    235 
    236 def hstack(tup):

ValueError: all the input arrays must have same number of dimensions
```